src/perform_step/bdf_perform_step.jl

```diff
@@ -262,20 +256,129 @@ function perform_step!(integrator,cache::QNDF1Cache,repeat_step=false)
   fail_convergence && return
   @. u = tmp + γ*z
 
-  if integrator.opts.adaptive
-    D2!(u,uprev,k,cache)
-    @. utilde = (κ*γₖ + inv(k+1)) * D2[1,1]
+  if integrator.opts.adaptive && integrator.success_iter > 0
+    @. D2[1] = u - uprev
+    @. D2[2] = D2[1] - D[1]
+    @. utilde = (κ*γ₁ + inv(k+1)) * D2[2]
     calculate_residuals!(atmp, utilde, uprev, u, integrator.opts.abstol, integrator.opts.reltol, integrator.opts.internalnorm)
     integrator.EEst = integrator.opts.internalnorm(atmp)
     if integrator.EEst > one(integrator.EEst)
-      cache.D .= temp_D
-      cache.uprev2 .= temp_u
       return
     end
+  else
+    integrator.EEst = one(integrator.EEst)
   end
   cache.dtₙ₋₁ = dt
   cache.uprev2 .= uprev
   cache.ηold = η
   cache.newton_iters = iter
   f(integrator.fsallast, u, p, t+dt)
 end
+
+function initialize!(integrator, cache::QNDF2ConstantCache)
+  integrator.kshortsize = 2
+  integrator.k = typeof(integrator.k)(integrator.kshortsize)
+  integrator.fsalfirst = integrator.f(integrator.uprev, integrator.p, integrator.t) # Pre-start fsal
+
+  # Avoid undefined entries if k is an array of arrays
+  integrator.fsallast = zero(integrator.fsalfirst)
+  integrator.k[1] = integrator.fsalfirst
+  integrator.k[2] = integrator.fsallast
+end
+
+function perform_step!(integrator,cache::QNDF2ConstantCache,repeat_step=false)
+  @unpack t,dt,uprev,u,f,p = integrator
+  @unpack pass1,pass2,pdt,uprev2,uprev3,dtₙ₋₁,dtₙ₋₂,D,D2,R,U = cache
+  cnt = integrator.iter
+  pdt = dt
+  k = 2
+  flag = dtₙ₋₁ != dtₙ₋₂
+
+  if flag && cnt != 1
+    if pass1
+      integrator.dt = dtₙ₋₁
+    elseif pass2
+      integrator.dt = pdt
+    end
+  end
+
+  dt = integrator.dt
+
+  if cnt == 1 || cnt == 2 || flag
+    κ = zero(integrator.alg.kappa)
+    γ₁ = 1//1
+    γ₂ = 1//1
+  else
+    κ = integrator.alg.kappa
+    γ₁ = 1//1
+    γ₂ = 1//1 + 1//2
+    ρ = dt/dtₙ₋₁
+    # backward diff
+    D[1] = uprev - uprev2
+    D[2] = D[1] - (uprev2 - uprev3)
+    if ρ != 1
+      R!(k,ρ,cache)
+      cache.D .= D * (R * U)
+    end
+  end
+
+  # precalculations
+  γ = inv((1-κ)*γ₂)
+  u₀ = uprev + D[1] + D[2]
+  ϕ = γ * (γ₁*D[1] + γ₂*D[2])
+  tmp = u₀ - ϕ
+
+  γdt = γ*dt
+  W = calc_W!(integrator, cache, γdt, repeat_step)
+
+  # initial guess
+  z = dt*integrator.fsalfirst
+
+  z, η, iter, fail_convergence = diffeq_nlsolve!(integrator, cache, W, z, tmp, γ, 1, Val{:newton})
+  fail_convergence && return
+  u = tmp + γ*z
+
+  if integrator.opts.adaptive && integrator.success_iter > 0 && flag
+    # bdf1_EEst!(integrator, u, uprev, cache)
+    D[1] = uprev - uprev2
```

Comment:
from line 343 to 349, I have implemented error est of `bdf1` from Matlab paper (formula no. 3) i.e `EEst = 1//2 * (u - 2*uprev + uprev2)` but when using this error est instead of error est of `ImplicitEuler` results to following error.

```
ERROR: Something went wrong. Integrator stepped past tstops but the algorithm was dtchangeable. Please report this error.
Stacktrace:
 [1] handle_tstop!(::OrdinaryDiffEq.ODEIntegrator{OrdinaryDiffEq.QNDF2{0,true,DiffEqBase.LinSolveFactorize{Base.LinAlg.#lufact!},DataType,Void,Void,Float64,Rational{Int64},:Predictive},Float64,Float64,Void,Float64,Float64,Float64,Array{Float64,1},DiffEqBase.ODESolution{Float64,1,Array{Float64,1},Void,Void,Array{Float64,1},Array{Array{Float64,1},1},DiffEqBase.ODEProblem{Float64,Float64,false,Void,#f,Void,Void,UniformScaling{Int64},DiffEqBase.StandardODEProblem},OrdinaryDiffEq.QNDF2{0,true,DiffEqBase.LinSolveFactorize{Base.LinAlg.#lufact!},DataType,Void,Void,Float64,Rational{Int64},:Predictive},OrdinaryDiffEq.InterpolationData{#f,Array{Float64,1},Array{Float64,1},Array{Array{Float64,1},1},OrdinaryDiffEq.QNDF2ConstantCache{DiffEqDiffTools.UDerivativeWrapper{#f,Float64,Void},Float64,Array{Float64,2},Array{Float64,2},Float64,Float64}}},#f,Void,OrdinaryDiffEq.QNDF2ConstantCache{DiffEqDiffTools.UDerivativeWrapper{#f,Float64,Void},Float64,Array{Float64,2},Array{Float64,2},Float64,Float64},OrdinaryDiffEq.DEOptions{Float64,Float64,Float64,Float64,DiffEqBase.#ODE_DEFAULT_NORM,DiffEqBase.CallbackSet{Tuple{},Tuple{}},DiffEqBase.#ODE_DEFAULT_ISOUTOFDOMAIN,DiffEqBase.#ODE_DEFAULT_PROG_MESSAGE,DiffEqBase.#ODE_DEFAULT_UNSTABLE_CHECK,DataStructures.BinaryHeap{Float64,DataStructures.LessThan},DataStructures.BinaryHeap{Float64,DataStructures.LessThan},Void,Void,Int64,Array{Float64,1},Array{Float64,1},Array{Float64,1}},Float64}) at /home/shubham00/.julia/v0.6/OrdinaryDiffEq.jl/src/integrators/integrator_utils.jl:417

```

Reply:
And the EEst values you were getting were sensible? You don't modify `dt` anywhere? This error wouldn't mean that this approach is incorrect or anything, it means there's some egregious `dt` modification going on.